### PR TITLE
Add smart decompositions to the `Exp` operator

### PR DESCRIPTION
**Pull request by AlbertMitjans**
## Add smart decompositions to the `Exp` operator
Decompose the `Exp` operator using the following rules:

1. If `Exp` contains the generator of another operator, return it when decomposing. 
2. If `Exp` contains a Pauli word, return a `PauliRot` operator. 
3. Return the Suzuki-Trotter decomposition. **This decomposition requires the `num_steps` attribute to be defined.**

**Note:** I added the decomposition to the `Exp` class to be more generic. This way we support the following decompositions:
```pycon
>>> qml.exp(1j * qml.PauliX(0), coeff=4).decomposition()
[RX((-8+0j), wires=[0])]
>>> qml.exp(1j * qml.sum(qml.PauliZ(1), qml.PauliY(0)), coeff=2, num_steps=1).decomposition()
[PauliRot((-4+0j), Z, wires=[1]), PauliRot((-4+0j), Y, wires=[0])]
```

**Note:** The `Evolution` class was added just to make `Exp` differentiable. In the long run I'd try to make `Exp` differentiable and remove the `Evolution` class, given that this class acts as a constructor of `Exp` adding `1j` to the coefficient (thus it can be just a function).

## Other improvements
For these changes to work correctly, other improvements have been made in several places of the PL codebase:

- **dot.py**: Cast `Tensor` to `Prod` and group coefficients by value.
- **generator.py**: Support `SProd` and `Sum` in `qml.generator`.
- **sum.py**: Also sort summands alphabetically.
- **parametric_ops.py** and **qchem_ops.py**: Simplify `generator` expressions.

### CODE REVIEW

**Comment by AlbertMitjans** [comment]
[sc-32752]

**Comment by albi3ro** [comment]
@AlbertMitjans  PR description please.

**Comment by AlbertMitjans** [comment]
> @AlbertMitjans PR description please.

This PR is still a work in progress. I added a general description. Will add a better one once the PR is ready for review.

File: pennylane/ops/op_math/exp.py
Comment (albi3ro) [inline pennylane/ops/op_math/exp.py:144]: Might be worth have a more expressive variable name.

File: pennylane/ops/op_math/exp.py
Comment (lillian542) [inline pennylane/ops/op_math/exp.py:53]:
```suggestion
        id (str): id for the Exp operator. Default is None.
```

File: pennylane/ops/op_math/evolution.py
Comment (lillian542) [inline pennylane/ops/op_math/evolution.py:40]:
```suggestion
        id (str): id for the Evolution operator. Default is None.
```

File: pennylane/ops/op_math/exp.py
Comment (lillian542) [inline pennylane/ops/op_math/exp.py:128]:
```suggestion
        id (str): id for the Exp operator. Default is None.
```

File: pennylane/ops/op_math/exp.py
Comment (mudit2812) [inline pennylane/ops/op_math/exp.py:236]: Maybe a good idea to add a shortcut story for this TODO

**Review by mudit2812 — COMMENTED** [review]
Looks good! I added some minor comments but overall it looks good to me! I think another reviewer should approve for this PR to get merged.

File: tests/ops/functions/test_dot.py
Comment (lillian542) [inline tests/ops/functions/test_dot.py:72]:
```suggestion
        assert isinstance(result.base, Sum)
        assert len(result.base) == 3
        for op in result.base:
```

File: tests/ops/op_math/test_exp.py
Comment (lillian542) [inline tests/ops/op_math/test_exp.py:393]: Since it doesn't have a decomposition, `num_steps` will never be used by this operator, right?

File: pennylane/ops/functions/generator.py
Comment (lillian542) [inline pennylane/ops/functions/generator.py:87]: I don't think I've seen this before, what does this do?

**Comment by lillian542** [comment]
Looks great! 🚀 I left a few very minor comments and questions :)

File: pennylane/ops/functions/generator.py
Comment (albi3ro) [inline pennylane/ops/functions/generator.py:119]: Just letting you know that the below code for `if op.inverse` is about to get deleted, see #3725 .

Once that code is removed, we could re-write this to use inner `return` statements and only have to use `if` statements instead of `elif`.

For example:
```
        if qml.math.allequal(gen.coeffs[0], gen.coeffs):
             # case where the Hamiltonian coefficients are all the same
             return sum(gen.ops),  gen.coeffs[0]
         if qml.math.allequal(abs_coeffs[0], abs_coeffs):
             # absolute value of coefficients is the same
             return qml.dot(coeffs, gen.ops), abs_coeffs[0]
```
Might allow us to clean some of this up.

File: pennylane/ops/op_math/exp.py
Comment (albi3ro) [inline pennylane/ops/op_math/exp.py:180]: Do we need to dynamically recalculate this upon initialization? Is this to accomodate someone adding a new operator to `has_unitary_generator` after initializing PL?

And why doesn't PauliRot count?

File: pennylane/ops/op_math/exp.py
Comment (albi3ro) [reply pennylane/ops/op_math/exp.py:236]: +1.  TODO's rarely ever get done when placed in code.

File: pennylane/ops/op_math/exp.py
Comment (albi3ro) [inline pennylane/ops/op_math/exp.py:377]: Be careful about queuing.

File: pennylane/ops/op_math/exp.py
Comment (albi3ro) [inline pennylane/ops/op_math/exp.py:307]: I wonder if we should add a new mode to `qml.generator` to absorb this logic into the `qml.generator` function itself.

File: pennylane/ops/op_math/exp.py
Comment (albi3ro) [inline pennylane/ops/op_math/exp.py:296]: This and `self._ops_with_generator` seem to be the only instance uses in this method. I'm wondering if we can make these static methods or just extract them from the class completely into "helper" functions.

I don't even think it'd be that bad to just calculate `_ops_with_generator` when we need it instead of caching it on individual instances.

File: pennylane/ops/op_math/exp.py
Comment (albi3ro) [inline pennylane/ops/op_math/exp.py:318]: Given `PauliRot` can already decompose into individual rotation gates and its a simpler pathway, I wonder if its worth making this check first, and then checking for the more complicated multi-term generator cases.

File: pennylane/ops/functions/generator.py
Comment (mudit2812) [reply pennylane/ops/functions/generator.py:87]: It's a variable type hint. It doesn't restrict the type of the variable to what you're specifying, but your IDE can use the type hint to infer what attributes and methods are available for that variable. You can read more about it [here](https://peps.python.org/pep-0484/).

File: tests/ops/op_math/test_exp.py
Comment (AlbertMitjans) [reply tests/ops/op_math/test_exp.py:393]: Yes... this is not ideal. The best solution would be to use `num_steps` when calling `decomposition` (thus the `Device` should hold the information about `num_steps`). But this solution would require a lot more thought and work...

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:180]: > Do we need to dynamically recalculate this upon initialization?

Yes, otherwise we'd get circular import errors... Ideally we should use class types (instead of strings) in all PL attributes.

Edit: I just realised we can move this list inside the attributes file, and build it only once.

> And why doesn't PauliRot count?

Because the initialisation is different from other operations (it requires a string containing the pauli word). Would be nice to support operators too.

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:375]: I just realised that when multiplying a list by an integer, we get multiple references of the **same** object:
```pycon
>>> obj = [1, 2, 3]
>>> a = [obj] * 3
>>> a
[[1, 2, 3], [1, 2, 3], [1, 2, 3]]
>>> obj[1] = 5
>>> a
[[1, 5, 3], [1, 5, 3], [1, 5, 3]]
```
Thus we won't have any queuing issues here.

Do you think this will likely bring problems in the future? As long as we don't update classes dynamically...

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:316]: But then we would never get an `Ising` gate (we would always get a `PauliRot`). Do we always prefer `PauliRot`?

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py]: > Edit: I just realised we can move this list inside the attributes file, and build it only once.

Done!

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:294]: We also use `self._pauli_rot_decomposition` and `self._trotter_decomposition`. And in `self._trotter_decomposition` we use `self.num_steps`.

We can definitely make all of them static methods, and pass `num_steps` into `_recursive_decomposition`. But what do we gain from this?

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:305]: We definitely can, but this is outside the scope of this PR. I can create an issue if you want.

File: pennylane/ops/op_math/sum.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/sum.py]: Thanks for the catch! Updated docstring.

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:234]: Where should I add the ticket (in which epic)? @albi3ro

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py]: Edit: sorry I misunderstood what you were asking for

Moved this inside `_generator_prefactor`, such that now it always returns a `Sum` and `Prod` instead of a `Hamiltonian` and `Tensor`.

File: pennylane/ops/functions/generator.py
Comment (lillian542) [reply pennylane/ops/functions/generator.py]: Cool - didn't know this could be used outside of function definitions :)

**Review by mudit2812 — COMMENTED** [review]
Looks good! I think there are a few conversations that should be had before the PR is merged. @albi3ro brought up a lot of really good points that further discussion might be good for. But overall, this looks great! :)

File: pennylane/ops/functions/generator.py
Comment (AlbertMitjans) [reply pennylane/ops/functions/generator.py]: Changed to using return statements!

File: pennylane/ops/op_math/exp.py
Comment (albi3ro) [reply pennylane/ops/op_math/exp.py:234]: "Evolution operator has smart decompositions"?

File: pennylane/ops/op_math/exp.py
Comment (albi3ro) [reply pennylane/ops/op_math/exp.py:363]:
```
>>> op = qml.evolve(qml.PauliX(0) + qml.PauliY(1), -0.5, num_steps=2)
>>> op.decomposition()
[RX((-0.5+0j), wires=[0]),
 RY((-0.5+0j), wires=[1]),
 RX((-0.5+0j), wires=[0]),
 RY((-0.5+0j), wires=[1])]
>>> op.expand().circuit
[RX((-0.5+0j), wires=[0]), RY((-0.5+0j), wires=[1])]
```

So yes, this is a problem.

In the medium term, I want to get `op.expand()` and make `op.decomposition` more flexible, but we haven't done that yet.

File: pennylane/ops/op_math/exp.py
Comment (albi3ro) [reply pennylane/ops/op_math/exp.py:304]: On further thought, we should indeed prioritize those other gates over `PauliRot`.

File: pennylane/ops/op_math/exp.py
Comment (albi3ro) [reply pennylane/ops/op_math/exp.py:288]: Just my general preference to avoid having too much source code in the class itself.

File: pennylane/ops/qubit/attributes.py
Comment (albi3ro) [inline pennylane/ops/qubit/attributes.py:170]: This seems like a definite limitation of `Attributes`.

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:234]: Ok! It would be nice to have an epic for these TODOs though

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:368]: Thanks for noticing this! I didn't know we had to make sure ops were queued in the decomposition method... 😓 

Fixed 👍

File: pennylane/ops/qubit/attributes.py
Comment (AlbertMitjans) [reply pennylane/ops/qubit/attributes.py:170]: What do you mean exactly?

File: pennylane/ops/op_math/exp.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:234]: Added `TODOs` epic in Shortcut and created a story with this.

I don't want to add a story in `Evolution operator has smart decompositions` because this epic is linked to Productboard, and I want to mark it as done as soon as this PR is merged.

File: doc/releases/changelog-dev.md
Comment (albi3ro) [inline doc/releases/changelog-dev.md:415]: Is this a merge conflict?

File: doc/releases/changelog-dev.md
Comment (albi3ro) [inline doc/releases/changelog-dev.md:418]: Makes it sound like `qml.generator(Sum)` is supported, not that `qml.generator(op, format="prefactor") returns sum or product instances.

File: pennylane/ops/qubit/attributes.py
Comment (albi3ro) [reply pennylane/ops/qubit/attributes.py:170]: It'd be nice if we could just use the `has_unitary_generator` attribute, instead of having to maintain two lists. I'm just wondering what would be the effect if attributes stored a set of classes, instead of string names.

File: pennylane/ops/op_math/exp.py
Comment (albi3ro) [inline pennylane/ops/op_math/exp.py:354]:
```suggestion
            raise DecompositionUndefinedError(
```

File: doc/releases/changelog-dev.md
Comment (AlbertMitjans) [reply doc/releases/changelog-dev.md:415]: Nope, these features were needed for this PR to work. I didn't create a PR for each of those because they were only 4-5 lines each.

File: doc/releases/changelog-dev.md
Comment (AlbertMitjans) [reply doc/releases/changelog-dev.md]: Changed!

**Review by albi3ro — APPROVED** [review]
Looks great 🎉

**Pull request merged**
